10_rest/willshen8/pkg/puppy/rest.go

```diff
@@ -0,0 +1,100 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// RestHandler provides a puppy storer to store the puppies
+type RestHandler struct {
+	store Storer
+}
+
+// NewRestHandler is a factory method that makes new storer
+func NewRestHandler(storer Storer) *RestHandler {
+	return &RestHandler{store: storer}
+}
+
+// HandleGet gets the puppy by id and displays the results.
+func (rh *RestHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvalidInput.String(), 400)
+	} else {
+		puppy, readErr := rh.store.ReadPuppy(uint32(id))
+		if readErr != nil {
+			w.WriteHeader(http.StatusNotFound)
+			http.Error(w, http.StatusText(404)+": "+ErrorReadPuppyValue.String(), 404)
+		} else {
+			render.JSON(w, r, puppy)
+		}
+		return
+	}
+}
+
+// HandlePost post a new puppy and then display the results.
+func (rh *RestHandler) HandlePost(w http.ResponseWriter, r *http.Request) {
+	var p Puppy
+	if decodeErr := render.DecodeJSON(r.Body, &p); decodeErr != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrornvalidJSONInput.String(), 400)
+	} else {
+		_, createErr := rh.store.CreatePuppy(&p)
+		if createErr != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			render.JSON(w, r, createErr)
+		} else {
+			render.JSON(w, r, p)
+		}
+		return
+	}
+}
+
+// HandlePut updates the existing puppy with new fields in the payload.
+func (rh *RestHandler) HandlePut(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		http.Error(w, http.StatusText(400)+": "+ErrorInvliadIDForUpdate.String(), 400)
+	} else {
+		var p Puppy
+		result, updateErr := rh.store.UpdatePuppy(uint32(id), &p)
```

Comment:
You are updating with uninitialized `p Puppy`. 
Sounds like you forgot:
```
 render.DecodeJSON(r.Body, &p)
```

```
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10   curl localhost:8888/api/puppy/8 -X PUT -d '{"value":"333"}'
true
(⎈ minikube:default)
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10   curl localhost:8888/api/puppy/8
{"id":8,"breed":"","color":""}

```

Reply:
I may be wrong, but seems like your update method doesn't check if `puppy` value is greater than 0. n this case I can crate puppy with value `42` and than update it to value `-42` which is against to specification.

In this case:
1. You should fix the update method.
2. Update can return error NotFound and invalidValue Err
3. This funtion should deal with update error(400/422) and not found err(404)

Reply:
It does cater for invalid ID:

Request: `http://localhost:8888/api/puppy/-42`
Response: `Not Found: Puppy ID can not be found, update operation failed.`

Reply:
Thanks for picking up the POST method, I accidentally deleted a line when I was refactoring the Errors.

Reply:
Hey, still doesn't work as expected:
```
curl localhost:8888/api/puppy/3 -X PUT -d 'I AM NOT A JSON' -v
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8888 (#0)
> PUT /api/puppy/3 HTTP/1.1
> Host: localhost:8888
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Length: 15
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 15 out of 15 bytes
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< Date: Sun, 25 Aug 2019 10:19:44 GMT
< Content-Length: 5
<
true
* Connection #0 to host localhost left intact
```

and

```
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl localhost:8888/api/puppy/4 -X PUT -d '{"value":"-333"}'
true
(⎈ minikube:default)
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl localhost:8888/api/puppy/4
{"id":4,"breed":"","color":"","value":"-333"}

```

I'm using server as:
```go run cmd/puppy-server/main.go -s map -d data/puppies.json```

Reply:
Now works, GOLANGCI picked up too... :) 
`Unprocessable Entity: Invalid JSON, please make sure the JSON puppy is valid.`

Reply:
Still not work as expected. Please refer to my 2nd comment for explanation what is wrong.

Please check following output:
```
 n0npax@narrenturm  /tmp/go-course/10_rest   lab10 ●   curl localhost:8888/api/puppy/4 -X PUT -d '{"value":"-777"}' -v
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8888 (#0)
> PUT /api/puppy/4 HTTP/1.1
> Host: localhost:8888
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Length: 16
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 16 out of 16 bytes
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< Date: Sun, 25 Aug 2019 12:13:48 GMT
< Content-Length: 5
<
true
* Connection #0 to host localhost left intact
(⎈ minikube:default)
 n0npax@narrenturm  /tmp/go-course/10_rest   lab10 ●   curl localhost:8888/api/puppy/4
{"id":4,"breed":"","color":"","value":"-777"}
```

Reply:
Another, less important issue:

I believe 1st request in given snippet should return `404`. In my mind It's not `422`. Request is valid, but resource doesn't exist.
```
 n0npax@narrenturm  /tmp/go-course/10_rest   lab10 ●   curl localhost:8888/api/puppy/114 -X PUT -d '{}' -i
HTTP/1.1 422 Unprocessable Entity
Date: Sun, 25 Aug 2019 12:21:42 GMT
Content-Length: 74
Content-Type: text/plain; charset=utf-8

Unprocessable Entity: Puppy ID can not be found, update operation failed.
(⎈ minikube:default)
 n0npax@narrenturm  /tmp/go-course/10_rest   lab10 ●   curl localhost:8888/api/puppy/4 -X PUT -d '{}' -i
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Sun, 25 Aug 2019 12:21:49 GMT
Content-Length: 5

true
```

Reply:
@willshen8 If given snippets are not clear or you need some additional explanation please reach me in the office. I'm not best in explanations via GH.